tcp_sink.go

```diff
@@ -4,105 +4,148 @@ import (
 	"bufio"
 	"fmt"
 	"net"
-	"sync/atomic"
+	"sync"
 	"time"
 
 	logger "github.com/sirupsen/logrus"
 )
 
 const (
-	logOnEveryNDropped = 1000
+	flushInterval           = time.Second
+	logOnEveryNDroppedBytes = 1 << 15 // Log once per 32kb of dropped stats
 )
 
-// NewTCPStatsdSink returns a Sink that is backed by a go channel with a limit of 1000 messages.
+// NewTCPStatsdSink returns a Sink that is backed by a buffered writer and a separate goroutine
+// that flushes those buffers to a statsd connection.
 func NewTCPStatsdSink() Sink {
-	sink := &tcpStatsdSink{
-		outc: make(chan string, 1000),
+	outc := make(chan []byte, 1000)
+	writer := sinkWriter{
+		outc: outc,
 	}
-	go sink.run()
-	return sink
+	bufWriter := bufio.NewWriter(writer)
+	bufPool := &sync.Pool{
+		New: func() interface{} {
+			return make([]byte, bufWriter.Size())
+		},
+	}
+	s := &tcpStatsdSink{
+		outc:      outc,
+		bufWriter: bufWriter,
+		bufPool:   bufPool,
+	}
+	writer.bufPool = s.bufPool
+	go s.run()
+	return s
 }
 
 type tcpStatsdSink struct {
-	conn            net.Conn
-	outc            chan string
-	droppedTimers   uint64
-	droppedCounters uint64
-	droppedGauges   uint64
+	conn         net.Conn
+	outc         chan []byte
+	droppedBytes uint64
+	mu           sync.Mutex
+	bufWriter    *bufio.Writer
+	bufPool      *sync.Pool
 }
 
-func (s *tcpStatsdSink) FlushCounter(name string, value uint64) {
+type sinkWriter struct {
+	bufPool *sync.Pool
+	outc    chan<- []byte
+}
+
+func (w sinkWriter) Write(p []byte) (int, error) {
+	n := len(p)
+	pCopy := w.bufPool.Get().([]byte)
+	if copy(pCopy, p) != len(p) {
+		panic("didn't copy all bytes! bufPool buffers are the wrong length")
```

Comment:
i'd feel comfortable panic'ing here iff we have a unit test to catch regressions

Reply:
I'm not cool with a panic at all in a library. Usually you keep a pool of `bytes.Buffer` instead of  just `[]byte` to avoid this.